Add Language Library Design Principles (#84)
open-telemetry/opentelemetry-specification: specification/library-guidelines.md

# OpenTelemetry Language Library Design Principles

> This document defines common principles that will help designers to create language libraries that are ergonomic to use, enable multiple use cases for library users, are to a certain extent uniform across all supported languages, yet allow enough flexibility for language-specific expressiveness.

Reviewer: what ergonomic mean? I'd suggest to use "easy" instead.

> This document defines common principles that will help designers to create language libraries that are ergonomic to use, enable multiple use cases for library users, are to a certain extent uniform across all supported languages, yet allow enough flexibility for language-specific expressiveness.

Reviewer: Remove " to a certain extent". We want them to be unifiorm with the caveat explained in the next part of a sentence

> This document defines common principles that will help designers to create language libraries that are ergonomic to use, enable multiple use cases for library users, are to a certain extent uniform across all supported languages, yet allow enough flexibility for language-specific expressiveness.

Reviewer: Also I'd say "full features out of the box and allow for innovation and experimentation thru extensibility poiunts"

> The document does not attempt to describe a language library API. For API specs see [Tracing API](tracing-api.md), [Resources API](resources-api.md).

Reviewer: ```suggestion
The document does not attempt to describe a language library API. For API specs see [specification](specification/README.md).
```

## Requirements

> 1. The public API must be well-defined and clearly decoupled from the implementation, enabling end-users to swap out for alternate implementations.

Reviewer: I'd also mention the package structure before going into requirements for individual layers. @mayurkale22 is working on packages document in #52 so there will be a document to reference to.  The most important thing why I think we need to start with packaging is that we need to clearly separate "public API" as a language concept and "public API" as an API for instrumentation points. Maybe even call it differently. Like "OpenTelemetry API"

There will be also "public API" in SDK that will allow to create exporters.

Author: I am not sure I understand. Can you clarify what do you mean by "package structure"? Are you referring to the fact that we want the language library to be composed of 2 packages: one providing the Public API and the other its implementation?

Reviewer: Yes. "Public API" is very confusing term before you'd explain it. Exporter interface is also public API. But it is not what we include into OpenTelemetry API package, it belong to SDK.

Author: OK, makes sense. Let me re-word.

3. The developers of the final application normally decide what OpenTelemetry implementation to plug in. They should be also free to choose to not use any OpenTelemetry implementation at all, even though the application and/or its libraries are already instrumented.  The rationale is that third-party libraries and frameworks which are instrumented with OpenTelemetry must still be fully usable in the applications which do not want to use OpenTelemetry (so this removes the need for framework developers to have "instrumented" and "non-instrumented" versions of their framework).

4. Full implementation must be clearly separated into wire protocol-independent parts that implement common logic (e.g. batching, tag enrichment by process information, etc.) and protocol-dependent exporters. Protocol-dependent exporters must contain minimal functionality, thus enabling vendors to easily add support for their specific protocol to the language library.

## Proposed Generic Design

Here is a proposed generic design for a language library:

> ![Language Library Design Diagram](language-library-design.png)

Reviewer: we call no-op implementation "Default" as it's not a no-op any longer. It propagates context

Author: Does "Default" convey the right meaning? People may confuse the "Default" with the default full SDK. 

I agree that no-op is also not a very term. 

Perhaps "Minimal" would be better?

Reviewer 2: @SergeyKanzhelev how do we distinguish between "default" (context propagation) and "default" (full standard implementation)?

> ![Language Library Design Diagram](language-library-design.png)

Reviewer: Please make is clear that end-user app take a dependency on SDK. From the picture one may think that API takes a dependency to the SDK

Author: The arrows in the picture are call graphs, not dependencies. There is no direct call from end-user app to SDK, so I do not want to add an arrow. Would it be sufficient if I add this explanation in the the text?

> ![Language Library Design Diagram](language-library-design.png)

Reviewer: Why protocol exporters are in Full SDK box? Wouldn't they be in separate vendor-specific box? Also "protocol exporter" is not a great name as may be confused with wire context propagation format. Maybe "telemetry exporter".

Author: Do we intend to include protocol exporters by default in the Full SDK or it will be up to the application developer to link with certain exporters? Looks like you are saying it is the later. If that's the case then you are right the diagram is not correctly depicting it.

Reviewer 2: My understanding is that we'll include Zipkin, Jaeger, and Prometheus out of the box ... all others will need to be linked

### Public API

The instrumentation code in the end-user application and in the instrumented third-party libraries that are used by the application will call the public API of OpenTelemetry Language Library.

The public API package is a self-sufficient dependency, in the sense that if the end-user application depends only on it and does not plug a full implementation then the application will still build and run without failing, although no telemetry data will be actually delivered to a telemetry backend.

This self-sufficiency is achieved the following way.

> The public API dependency will contain a minimal, virtually no-op implementation of the API. When no other implementation is explicitly included in the application all telemetry calls will effectively be no-op, any generated telemetry data will be discarded. 

Reviewer: I'd exclude all mentions of no-op to avoid confusion. Something like: 

```suggestion
The public API dependency will contain a minimal implementation of the API. When no other implementation is explicitly included in the application no telemetry data will be collected.
```

Author: Sounds good. Looks like using the word "minimal" is clear enough, let me fix that.

It is important that values returned from this no-op implementation of API are valid and do not require the caller to perform extra checks (e.g. createSpan() method should not fail and should return a valid non-null Span object). The caller should not need to know and worry about the fact that no-op implementation is in effect. This minimizes the boilerplate and error handling in the instrumented code.

It is also important that no-op implementation incurs as little performance penalty as possible, so that third-party frameworks and libraries that are instrumented with OpenTelemetry impose negligible overheads to users of such libraries that do not want to use OpenTelemetry too.

> ## Full SDK Implementation

Reviewer: is the word "Full" important? Kind of confusing. Is there non-full SDK that I can use?

Author: I used "Full" to emphasize that it is the opposite of "No-op". Do you think it is sufficient to simply refer to SDK? I am worried that SDK is a generic word that may not convey the message properly.

Reviewer: I'd suggest to call it OpenTelemetry SDK. A bit longer, but  doesn't raise question on where to get partial SDK =). Maybe earlier in the document - in packaging section you can give a definition

The Full SDK implementation is a separate (optional) dependency. When it is plugged in it substitutes the no-op implementation that is included in the Public API package (exact substitution mechanism is language dependent).

> It is recommended to decouple common parts of the implementation from protocol-specific portion. Library designers are encouraged to specify an internal Exporter API that defines how protocol-independent and protocol-dependant parts interact. The design goals for such API will be:

Reviewer: same comment about the protocol

> It is recommended to decouple common parts of the implementation from protocol-specific portion. Library designers are encouraged to specify an internal Exporter API that defines how protocol-independent and protocol-dependant parts interact. The design goals for such API will be:

Reviewer: Also SDK will provide a set of processors which implement common patterns like queue-ing and batching. I'd say that the design pattern we need to pursue with vendor-specific implementations is to make sure it's possible for a vendor to write a simple, virtually "synchronous" channel that receives the batch of telemetry and sends this telemetry to the destination. So "getting started" is easy. And than those processors can be replaced if alternative implementation for efficient batching is required.

Author: Sounds good. I'll add more details.

- Minimize burden of implementation for protocol-dependant parts and avoid duplicate implementation of the same functionality.

- Use efficient data structures that are well suited for fast serialization to wire formats and minimize the pressure on memory managers. The latter typically requires understanding of how to optimize the rapidly-generated, short-lived telemetry data structures to make life easier for the memory manager of the specific language. General recommendation is to minimize the number of allocations and use allocation arenas where possible, thus avoiding explosion of allocation/deallocation/collection operations in the presence of high rate of telemetry data generation.

## Alternate Implementations

> The end-user application may decide to take a dependency on alternate implementation. 

Reviewer: I'd start with "SDK provides flexibility and extensibility that may be used by many implementations. Before developing an alternative implementation, please, review extensbility points provided by OpenTelemetry."

> The alternate implementation may provide functionality that is identical to the official full SDK, but for example has better performance characteristics. This opens up the possibility of having competing and better implementations.

Reviewer: I think the main purpose of alternative implementaiton is to implement alternative data models. Like event-based rather than span-based model.

This sentence is bothersome as it reads like "we will not provide efficient implementation".

> Another common use case for alternate implementations is automated testing. A mock implementation can be plugged in during automated tests. For example it can store all generated telemetry data in memory and provide a capability to inspect this stored data. This will allow the tests to verify that the telemetry is generated correctly. Language Library authors are encouraged to provide such mock implementation.

Reviewer: we discussed that for mocking purposes it's better to use SDK with the mock exporter or processor. This way an actual logic will be tested. Should we encourage this use case at all?

Reviewer 3: +1 to mock exporter with SDK, this is the model we used in OT. That said, I think it'll be incumbent on anyone doing a replacement implementation to provide their own mock.

Author: I thought about making the mock just another exporter but I think there are certain benefits for replacing the entire SDK with a mock implementation.

The SDK as we envision it may be too thick for simple mocking. Why do I need batching, tagging and all things that SDK does by default in a mock? When I use a mock in my automated tests I do not intend to verify how SDK does its job, I intend to verify if my instrumented code is calling telemetry API correctly. All I care about in the mock is to be able to inspect the results of telemetry calls.

The SDK's internal logic can be a detriment to my testing goals. For example if it batches and reorganizes telemetry data it may make inspection of data more difficult in the tests. This unneeded logic will also impose a small performance penalty on automated tests.

I think there are probably valid use cases for both approaches. The mocking by replacing only the exporter can be better suited for broader automated tests that are closer to integration tests or end-to-end tests.

Reviewer 3: How much of an overhead do we really think there will be? I got my terms mixed up earlier - we provide a mock tracer in OT, rather than a mock exporter as there's no exporter concept there - but it seems like a mock exporter is going to be easier for third parties to implement. Moreover, it seems like it'd be valuable to be able to understand the performance overhead of the SDK on the application code.

Author: @austinlparker you are right that likely the performance overhead is going to be minimal. The unnecessary complication of tests may be non-trivial though.

Think about how you would test if the mock is a full SDK with a mock Exporter. 

Your test executes a function that calls telemetry API. This in turn ends up in the SDK which decides to do batching and does NOT call the mock Exporter immediately. The SDK calls the mock Exporter after some timeout that is part of batching logic.

Now your test has to asynchronously wait for SDK batching logic timeout. This complicates the test, makes it potentially much slower and also makes it be aware of internal implementation details of the SDK, particularly that there is a batching logic with timeout.

I can envision other troubles as well. E.g. what if batching also reorders data for efficiency purposes? Now I need special logic in my test to find the telemetry data that I am looking for rather than be sure that the first data I see is what I need.

# OpenTelemetry Language Library Design Principles

> This document defines common principles that will help designers to create language libraries that are easy to use, enable multiple use cases for library users, are uniform across all supported languages, yet allow enough flexibility for language-specific expressiveness. 

Reviewer 2: ```suggestion
This document defines common principles that will help designers create language libraries that are easy to use, are uniform across all supported languages, yet allow enough flexibility for language-specific expressiveness. 
```

The language libraries are expected to provide full features out of the box and allow for innovation and experimentation through extensibility points.

The document does not attempt to describe a language library API. For API specs see [specification](specification/README.md).

## Requirements

> 1. The OpenTelemetry API must be well-defined and clearly decoupled from the implementation, enabling end-users to swap out for alternate implementations.

Reviewer: sorry for bashing on it. I think the main goal is not to swap, but to minimize the number of dependencies and have a very stable package. Enabling to swap should not be the first requirement from my perspective

Author: Makes sense, let me change.

Author: Fixed.

2. Third party libraries and frameworks that add instrumentation to their code will have a dependency only on the API of OpenTelemetry language library. The developers of third party libraries and frameworks do not care (and cannot know) what specific implementation of OpenTelemetry is used in the final application.

> 3. The developers of the final application normally decide what OpenTelemetry implementation to plug in. They should be also free to choose to not use any OpenTelemetry implementation at all, even though the application and/or its libraries are already instrumented.  The rationale is that third-party libraries and frameworks which are instrumented with OpenTelemetry must still be fully usable in the applications which do not want to use OpenTelemetry (so this removes the need for framework developers to have "instrumented" and "non-instrumented" versions of their framework).

Reviewer: I'd swap "decide what OpenTelemetry implementation to plug in." to "how to configure OpenTelemetry SDK and what extensions to use."

Again, the main use case and the project aspiration that there is a single implementation in most cases and we all generally aligned on data formats. The greatest outcome if every app reports telemetry in OpenTelemetry Agent format and only Agent and Collector are being configured to process and upload data to the proper location

Author: Fixed.

## Language Library Generic Design

Here is a generic design for a language library (arrows indicate calls):

![Language Library Design Diagram](language-library-design.png)

### Expected Usage

The OpenTelemetry Language Library will be composed of 2 packages: API package and SDK package.

Third-party libraries and frameworks that want to be instrumented in OpenTelemetry-compatible way will have a dependency on the API package. The developers of these third-party libraries will add calls to telemetry API to produce telemetry data.

Applications that use third-party libraries that are instrumented with OpenTelemetry API will have a choice to enable or not enable the actual delivery of telemetry data. 

In order to enable it the application must take a dependency on the OpenTelemetry SDK, which implements the delivery of the telemetry. The application must also configure exporters so that the SDK knows where and how to deliver the telemetry. The details of how exporters are enabled and configured are language specific.

> The application can also call telemetry API directly to produce additional telemetry data.

Reviewer 2: I'd recommend adding this to the text two paragraphs above, on line 32

It is also important that minimal implementation incurs as little performance penalty as possible, so that third-party frameworks and libraries that are instrumented with OpenTelemetry impose negligible overheads to users of such libraries that do not want to use OpenTelemetry too.

### SDK Implementation

The SDK implementation is a separate (optional) dependency. When it is plugged in it substitutes the minimal implementation that is included in the API package (exact substitution mechanism is language dependent). Here is what active components will look like in this case:

![Full Operation Diagram](language-library-full.png)

It is recommended to decouple common parts of the implementation from protocol-specific Telemetry Exporters. Library designers are encouraged to specify an internal Telemetry Exporter API that defines how protocol-independent and protocol-dependant parts interact. The boundary and design goals for such API will be:

> - Place common functionality such as queuing, batching, tagging, etc. in the protocol-independent parts of the implementation. This functionality will be applicable regardless of what Telemetry Exporter is used.

Reviewer 4: One thing we might want to consider here: how would people troubleshoot the SDK?
In most OpenCensus SDKs, we have `z-pages`. Also we've discussed that we might want to have metrics/logs for dropped events.

Some design principles we could explore:
1. The SDK/exporter should not block the customer's code (except for explicit call to `flush` and a grace exit), the SDK would rather drop data instead of eating up all the resources.
2. The SDK/exporter should have a confined resource consumption, e.g. max memory / max network bandwidth.

Author: Good points. There are probably more things that we would want to add to the design principles, so I suggest that we add it in a follow up PR to avoid blocking this one.

Author: I added an issue to track this: https://github.com/open-telemetry/opentelemetry-specification/issues/94

- Minimize burden of implementation for protocol-dependant telemetry exporters. The Telemetry Exporter is expected to be primarily a simple telemetry data encoder and transmitter.

- Use efficient data structures that are well suited for fast serialization to wire formats and minimize the pressure on memory managers. The latter typically requires understanding of how to optimize the rapidly-generated, short-lived telemetry data structures to make life easier for the memory manager of the specific language. General recommendation is to minimize the number of allocations and use allocation arenas where possible, thus avoiding explosion of allocation/deallocation/collection operations in the presence of high rate of telemetry data generation.

### Alternate Implementations

The end-user application may decide to take a dependency on alternate implementation. 

> SDK provides flexibility and extensibility that may be used by many implementations. Before developing an alternative implementation, please, review extensibility points provided by OpenTelemetry.

Reviewer 2: ```suggestion
The SDK provides flexibility and extensibility that may be used by many implementations. Before developing an alternative implementation, please, review extensibility points provided by OpenTelemetry.
```

> An example use case for alternate implementations is automated testing. A mock implementation can be plugged in during automated tests. For example it can store all generated telemetry data in memory and provide a capability to inspect this stored data. This will allow the tests to verify that the telemetry is generated correctly. Language Library authors are encouraged to provide such mock implementation.

Reviewer: I liked the mention you had before about alternative data models. Like event-based, not span-based backends. Alternative Tracer would be most useful in such examples.
